apps/src/lab2/lab2Redux.ts

```diff
@@ -524,6 +526,31 @@ async function cleanUpProjectManager() {
   Lab2Registry.getInstance().clearProjectManager();
 }
 
+// Returns if the current state represents a predict level taht should be read only.
+// If the predict level code is not editable after submit or the user has not submitted a response,
+// the predict level is read only.
+function isReadonlyPredictLevel(state: RootState) {
+  const isPredictLevel =
+    state.lab.levelProperties?.predictSettings?.isPredictLevel || false;
+  let isReadonlyPredictLevel = isPredictLevel;
+  if (isPredictLevel) {
+    const isEditableAfterSubmit =
+      state.lab.levelProperties?.predictSettings?.codeEditableAfterSubmit ||
+      false;
+    const hasSubmittedPredictResponse = state.predictLevel.hasSubmittedResponse;
+    // If the predict level code is not editable after submit or the user has not submitted a response,
+    // the predict level is read only.
+    isReadonlyPredictLevel =
+      !isEditableAfterSubmit || !hasSubmittedPredictResponse;
+  }
+  return isReadonlyPredictLevel;
+}
+
+// Currently only Python Lab disables editing while code is running.
+function shouldBeReadonlyWhileRunning(state: RootState) {
+  return state.lab.levelProperties?.appName === 'pythonlab';
```

Comment:
I'm not sure there's a better way to do this than to have this checked here for now (I'm curious what others think), and I'd expect there might be other labs that use this in the future.